sdk/common/src/main/java/io/opentelemetry/sdk/internal/ThrottlingLogger.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.internal;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import io.opentelemetry.sdk.common.Clock;
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.IntStream;
+import javax.annotation.Nullable;
+
+/** Will limit the number of log messages emitted, so as not to spam when problems are happening. */
+public class ThrottlingLogger {
+  private static final int RATE_LIMIT = 5;
+  private static final int THROTTLED_RATE_LIMIT = 1;
+  private static final TimeUnit rateTimeUnit = MINUTES;
+
+  private final Logger delegate;
+  private final Clock clock;
+
+  // this window tracks the number of logs messages per time unit. It holds the times of `log`
+  // calls.
+  private final Deque<Long> window = new ConcurrentLinkedDeque<>();
```

Comment:
This looks something like a leaky bucket sampler. Can we just move the rate limiter from jaeger remote sampler package and use it? It doesn't need a queue.

Reply:
The rate limiter from the jaeger sampler has a bit of a different goal. This isn't really a rate limiter that I've implemented. Instead, it's a throttle that gets invoked if the rate exceeds a value. The throttle is never released, though, unlike the Jaeger rate limiter which will let a steady number through for all time.

My goal here was to have an exporter log noisily initially (up to the 5 per minute limit), but if it's a known noisy logger, then have it just be clamped down to 1 per minute after that. That's a bit different than a rate limiter.

Reply:
I think the jaeger rate limiter, since its leaky bucket, is similar - if maxBalance is relatively hard and creditsPerSecond is set to one per minute, it would allow a lot of messages at first and then switch to 1 per minute

https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/RateLimiter.java#L21

At least I think so :)

The main difference is that the jaeger limiter could recover if the exporter stops being noisy. But that's ok I guess?

Reply:
Yeah, I wasn't sure if it was ok or not. If you've got a known-noisy logger, do you really want to let it back in to noisy mode?

 I personally think the queue-based implementation is fairly elegant, as you can simply adjust the length of the queue in order to raise or lower the current limit. But, I'll take another look at that RateLimiter and see if I can generate similar behavior to this one.

Reply:
But I don't like math :P So for the current throttled_rate_limit of 1 per minute, `creditsPerSecond = 1 / 60`. Then just need to decide how many log messages to allow before throttling, `maxBalance = 5` would log 5 consecutive messages without throttling and messages after that would be throttled to once per minute - unless logs stopped happening in which case after 5 minutes the rate limiter would be back to the initial state.

This might not be exactly what you have here but it seems like a reasonable strategy and I think it's close. The queue seems like it could be some decent overhead for something we want to actually be a no-op

Reply:
I'm definitely not worried about overhead in the logging wrapper. The whole point of this thing is to reduce the amount of logging which is happening on the exporter (background) threads. The overhead is really irrelevant here, IMO.

Reply:
ok, I re-implemented, using the Jaeger RateLimiter, and moved that class into the sdk-common internal package.